src/game/common/rts/productionprerequisite.cpp

```diff
@@ -50,3 +52,139 @@ int ProductionPrerequisite::Get_All_Possible_Build_Facility_Templates(ThingTempl
     }
     return i;
 }
+
+int ProductionPrerequisite::Calc_Num_Prereq_Units_Owned(const Player *player, int *counts) const
+{
+    unsigned int size = m_prereqUnit.size();
+
+    if (size > 32) {
+        size = 32;
+    }
+
+    ThingTemplate *thing[32];
+
+    for (int i = 0; i < size; i++) {
+        thing[i] = m_prereqUnit[i].unit;
+    }
+
+    player->Count_Objects_By_ThingTemplate(size, thing, false, counts, true);
+    return size;
+}
+
+bool ProductionPrerequisite::Is_Satisifed(const Player *player) const
+{
+    if (player == nullptr) {
+        return false;
+    }
+
+    for (unsigned int i = 0; i < m_prereqScience.size(); i++) {
+        if (!player->Has_Science(m_prereqScience[i])) {
+            return false;
+        }
+    }
+
+    int counts[32];
+    int num = Calc_Num_Prereq_Units_Owned(player, counts);
+
+    for (int i = 1; i < num; i++) {
+        if ((m_prereqUnit[i].flags & 1) != 0) {
+            counts[i] += counts[i - 1];
+            counts[i - 1] = -1;
+        }
+    }
+
+    for (int i = 1; i < num; i++) {
+        if (counts[i] != -1 && counts[i] == 0) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+Utf16String ProductionPrerequisite::Get_Requires_List(const Player *player) const
+{
+    if (player != nullptr) {
+        Utf16String str(Utf16String::s_emptyString);
+        int counts[32];
```

Comment:
This `32` value is repeated several times in these functions. What does this constant refer to? Is there a class member which shares this 32? Can we have a named constant for this or `ARRAY_SIZE(something)` ?